## <center> Assignment 2 </center>

#### Name: Kaniz Fatema
#### Student ID: 245832180

In [227]:
# Import all the essential libraries
import pandas as pd
import seaborn as sns
from matplotlib import pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, matthews_corrcoef, f1_score

In [228]:
# Load the datasets
train_data = pd.read_csv("train.csv")
test_data = pd.read_csv("test_attribute.csv")

In [229]:
# Understanding the given training dataset
train_data_columns = ["0", "1", "2", "3", "4", "5", "6", "7", "8"]

# Calculate and print the frequency of each value in the specified columns
for column in train_data_columns:
    if column in train_data.columns:  # Check if the column exists in train_data
        print(f"\033[1m\nFrequency for column {column}:\033[0m")
        frequency = train_data[column].value_counts()
        frequency.index.name = None  # Remove index name to avoid unexpected output
        frequency.name = None        # Remove series name to avoid 'Name: count' in output
        print(frequency)
    else:
        print(f"\033[1;31mColumn '{column}' not found in train_data.\033[0m")



Frequency for column 0:
0.51    32
0.46    31
0.45    29
0.50    25
0.47    23
        ..
0.83     1
0.19     1
0.22     1
0.20     1
0.24     1
Length: 70, dtype: int64

Frequency for column 1:
0.46    32
0.45    32
0.48    28
0.53    24
0.51    23
        ..
0.21     1
0.14     1
0.18     1
1.00     1
0.80     1
Length: 68, dtype: int64

Frequency for column 2:
0.54    42
0.53    40
0.51    39
0.52    37
0.50    35
0.55    31
0.56    29
0.49    29
0.48    26
0.47    25
0.45    24
0.57    23
0.46    22
0.58    20
0.43    15
0.35    14
0.38    14
0.36    14
0.34    14
0.44    13
0.42    13
0.59    11
0.60    10
0.32     9
0.41     9
0.30     9
0.61     8
0.62     8
0.33     8
0.63     7
0.64     7
0.40     7
0.39     7
0.37     5
0.29     3
0.67     3
0.71     3
0.31     2
0.75     2
0.26     2
0.65     2
0.21     2
0.69     2
0.27     1
0.24     1
0.66     1
1.00     1
0.22     1
0.72     1
0.28     1
dtype: int64

Frequency for column 3:
0.18    38
0.17    32
0.19    32
0.16    30
0

In [230]:
# Identify and report the number of missing values in the training dataset.
missing_values_count = train_data.isna().sum()
# Report the number of missing values in each column
print("\033[1mNumber of missing values in each column:\033[0m")
print(missing_values_count)
# Report the total number of missing values in the dataset
total_missing_values = missing_values_count.sum()
# Print the total number of missing values in bold
print(f"\033[1m\nTotal number of missing values in the crx dataset: {total_missing_values}\033[0m")

Number of missing values in each column:
Unnamed: 0    0
0             0
1             0
2             0
3             0
4             0
5             0
6             0
7             0
8             0
dtype: int64

Total number of missing values in the crx dataset: 0


In [231]:
# Separate features and labels
x = train_data.iloc[:, :-1]
y = train_data.iloc[:, -1]

In [232]:
# Print the values of x and y
print("\033[1mIllustration of the X values:\033[0m")
print("\033[1m---------------------------------------\033[0m")
print(x.head())

print("\033[1mIllustration of the Y values:\033[0m")
print("\033[1m---------------------------------------\033[0m")
print(y.head())

Illustration of the X values:
---------------------------------------
   Unnamed: 0     0     1     2     3    4    5     6     7
0           0  0.81  0.85  0.47  0.37  0.5  0.0  0.56  0.22
1           1  0.70  0.58  0.53  0.39  0.5  0.0  0.59  0.22
2           2  0.72  0.73  0.41  0.28  0.5  0.0  0.44  0.22
3           3  0.78  0.69  0.44  0.26  0.5  0.0  0.54  0.22
4           4  0.74  0.82  0.46  0.24  0.5  0.0  0.48  0.22
Illustration of the Y values:
---------------------------------------
0    1
1    1
2    1
3    1
4    1
Name: 8, dtype: int64


In [233]:
# Split training data for testing
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=1234)
print(f"\033[1mTraining set shape: {x_train.shape}, Testing set shape: {x_test.shape}\033[0m")

Training set shape: (521, 9), Testing set shape: (131, 9)


In [234]:
# Data Processing: Scale the data
scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [235]:
# Count number of 0s and 1s in training and testing labels
train_label_counts = y_train.value_counts()
test_label_counts = y_test.value_counts()
print("\033[1mCounts of labels in the Training set:\033[0m")
print(train_label_counts)
print("\033[1mCounts of labels in the Testing set:\033[0m")
print(test_label_counts)

Counts of labels in the Training set:
8
0    465
1     56
Name: count, dtype: int64
Counts of labels in the Testing set:
8
0    123
1      8
Name: count, dtype: int64


### Classifier-1: KNeighborsClassifier (KNN) .

In [224]:
# KNeighborsClassifier (KNN)
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(x_train, y_train)

# Make predictions on the training and testing sets
y_train_pred = knn.predict(x_train)
y_test_pred = knn.predict(x_test)

# Testing set performance
knn_test_accuracy = accuracy_score(y_test, y_test_pred)
knn_test_mcc = matthews_corrcoef(y_train, y_train_pred)
knn_test_f1 = f1_score(y_test, y_test_pred, average='weighted')

print('\033[1mModel performance for Testing set using KNN\033[0m')
print('- Accuracy: %s' % knn_test_accuracy)
print('- MCC: %s' % knn_test_mcc)
print('- F1 score: %s' % knn_test_f1)


Model performance for Testing set using KNN
- Accuracy: 0.9694656488549618
- MCC: 0.9284521016642407
- F1 score: 0.9710368609129716


### Generating the prediction.txt file for the KNeighborsClassifier (KNN) classifier.

In [225]:
# Scale the test dataset and make predictions for the KNN model
x_test_scaled = scaler.transform(test_data)
y_test_pred = knn.predict(x_test_scaled)
# Save predictions to "prediction_KNN.txt"
pd.DataFrame(y_test_pred, columns=["Predicted"]).to_csv("prediction_KNN.txt", header=False, index=False)
print("\033[1mPredictions saved to prediction_KNN.txt\033[0m")

Predictions saved to prediction_KNN.txt


### Classifier-2: Logistic Regression Classifier.

In [226]:
# Logistic Regression Classifier
from sklearn.linear_model import LogisticRegression 
log_reg = LogisticRegression(random_state=42)  # Instantiate Logistic Regression
log_reg.fit(x_train, y_train)  # Fit the model

# Make predictions on the training and testing sets
y_train_pred = log_reg.predict(x_train)
y_test_pred = log_reg.predict(x_test)

# Testing set performance
log_reg_train_accuracy = accuracy_score(y_test, y_test_pred)
log_reg_train_mcc = matthews_corrcoef(y_test, y_test_pred)
log_reg_train_f1 = f1_score(y_test, y_test_pred, average='weighted')

print('\033[1mModel performance for Testing set using Logistic Regression\033[0m')
print('- Accuracy: %s' % log_reg_test_accuracy)
print('- MCC: %s' % log_reg_test_mcc)
print('- F1 score: %s' % log_reg_test_f1)

Model performance for Testing set using Logistic Regression
- Accuracy: 0.9847328244274809
- MCC: 0.866869918699187
- F1 score: 0.9847328244274809


### Generating the prediction.txt file for the Logistic Regression (LR) classifier.

In [187]:
# Scale the test dataset and make predictions for the LR model
x_test_scaled = scaler.transform(test_data)
y_test_pred = log_reg.predict(x_test_scaled)
# Save predictions to "prediction_LR.txt"
pd.DataFrame(y_test_pred, columns=["Predicted"]).to_csv("prediction_LR.txt", header=False, index=False)
print("\033[1mPredictions saved to prediction_LR.txt\033[0m")

Predictions saved to prediction_LR.txt


### Briefly describe your approach in the following cell.

# Here's a brief description of the approach in the provided code cell: ( All the process is occured in a sequential order)
1. Library Importation: Import pandas, seaborn, matplotlib.pyplot, and necessary sklearn classes for data processing, visualization, and assessment.
2. Data Loading: Training and test datasets are loaded into DataFrames with pandas. This allows for the following analysis and processing of the datasets.
3. Data understanding (Finding frequencies and missing value analysis):
   Finding Frequency: The value_counts() function is used to compute and report the frequency of values in provided columns. Column existence is checked to prevent problems. The frequency Series index and name are reset to improve output quality.

   Missing Value Analysis: The amount of missing values in each column is calculated and reported. This contains both the number of missing values per column and the overall number of missing values in the dataset, which are bolded for emphasis.

4. Modeling involves separating characteristics (X) and labels (y). The characteristics include all columns except the last one, with the last column considered to represent the label of the dataset.

5. Data Inspection: Print the first rows of the feature set (X) and label set (y) to get an overview of the data structure and content.

6. Data Splitting: To guarantee repeatability, the dataset is split into training and testing subsets using sklearn's train_test_split function. The test size and random state are supplied by examining with several parameters.

7. Data Scaling: StandardScaler is used to normalize feature values, which is important for machine learning methods.

8. Label Count Analysis: We compute and report the counts of labels (0s and 1s) in both training and testing sets to get insight into class distribution.

# Classifier: After using the methods described above, it is critical to choose the best classifier to classify (Binary) the data. In this regard, numerous models were used. Since KNN and LR outperformed the others, these are finalized based on their performance. Furthermore, when compared to the KNN model (F1 score: 0.9710368609129716), the LR model outperformed (using testing data) it in terms of F1-score (F1 score: 0.9847328244274809).


10. # For the Optimal Model: Logistic Regression (LR) classifier.

Initiating the model: Sklearn.linear_model's LogisticRegression() instantiates a logistic regression model.
Model training: Fitting the model to x_train and y_train data. Training entails learning the feature-target variable connection.Training and testing datasets are predicted by the trained model. Result files are y_train_pred and y_test_pred.
Evaluate Performance: Calculate models' Accuracy,MCC, and F1 score on testing data set.

11. # Generating the prediction.txt file for the Logistic Regression (LR) classifier.
The scaler.transform method is used to standardize the test dataset, ensuring a mean of 0 and standard deviation of 1. The predicted labels are then generated using the trained logistic regression model ( utilizing the given testing dataset). The predictions are saved to a text file named "prediction_LR.txt" using a Pandas DataFrame. A confirmation message is then printed to the console.









